presto-hive/src/main/java/io/prestosql/plugin/hive/AcidInfo.java

```diff
@@ -0,0 +1,243 @@
+    {
+        return originalFileLocations;
+    }
+
+    @JsonProperty
+    public Optional<Integer> getBucketId()
+    {
+        return bucketId;
+    }
+
+    public boolean isSameBucket(Path path1, Path path2, Configuration conf)
+            throws IOException
+    {
+        if (bucketId.isPresent()) {
+            return bucketId.get() == AcidUtils.parseBaseOrDeltaBucketFilename(path2, conf).getBucketId();
+        }
+        return AcidUtils.parseBaseOrDeltaBucketFilename(path2, conf).getBucketId() ==
+                AcidUtils.parseBaseOrDeltaBucketFilename(path1, conf).getBucketId();
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        AcidInfo acidInfo = (AcidInfo) o;
+
+        if (!deleteDeltaLocations.equals(acidInfo.deleteDeltaLocations)) {
+            return false;
+        }
+        if (!originalFileLocations.equals(acidInfo.originalFileLocations)) {
+            return false;
+        }
+        return bucketId.equals(acidInfo.bucketId);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        int result = deleteDeltaLocations.hashCode();
+        result = 31 * result + originalFileLocations.hashCode();
+        result = 31 * result + bucketId.hashCode();
+        return result;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("deleteDeltaLocations", deleteDeltaLocations)
+                .add("originalFileLocations", originalFileLocations)
+                .add("bucketId", bucketId)
+                .toString();
+    }
+
+    public static Builder builder(Path partitionPath)
+    {
+        return new Builder(partitionPath);
+    }
+
+    public static class Builder
+    {
+        private Optional<DeleteDeltaLocations> deleteDeltaLocations;
+        private Optional<OriginalFileLocations> originalFileLocations;
+        private Optional<Integer> bucketId;
+        private final Map<Integer, List<OriginalFileInfo>> bucketIdToOriginalFileInfoMap = new HashMap<>();
+        private final Path partitionLocation;
+
+        private Function<FileStatus, Integer> bucketIdProvider;
+
+        private Builder(Path partitionLocation)
+        {
+            this.partitionLocation = requireNonNull(partitionLocation, "partitionLocation is null");
+            this.deleteDeltaLocations = Optional.empty();
+            this.originalFileLocations = Optional.empty();
+            this.bucketId = Optional.empty();
+        }
+
+        public Builder addOriginalFiles(Configuration configuration, List<HadoopShims.HdfsFileStatusWithId> originalFileLocations)
+        {
+            bucketIdProvider = fileStatus -> {
+                try {
+                    return AcidUtils.parseBaseOrDeltaBucketFilename(fileStatus.getPath(), configuration).getBucketId();
```

Comment:
We parse bucket numbers in `io.prestosql.plugin.hive.BackgroundHiveSplitLoader#getBucketNumber`